Generator/Sources/NeedleFramework/Generating/DependencyProviderSerializerTask.swift

```diff
@@ -34,28 +34,54 @@ class DependencyProviderSerializerTask: AbstractTask<[SerializedProvider]> {
     ///
     /// - returns: The list of `SerializedProvider`.
     override func execute() -> [SerializedProvider] {
-        return providers.map { (provider: ProcessedDependencyProvider) in
-            return serialize(provider)
+        var result = [SerializedProvider]()
+        // Group the providers based on where the properties are coming from
+        // This will allow us to extract common code for multiple depndency providers
+        // into common base classes
+        var counts = [[ProcessedProperty]: [ProcessedDependencyProvider]]()
+        for provider in providers {
+            let properties = provider.processedProperties
+            counts[properties, default: []].append(provider)
         }
+        for (_, (_, matchingProviders)) in counts.enumerated() {
+            result.append(contentsOf: serialize(matchingProviders))
+        }
+        return result
     }
 
     // MARK: - Private
 
     private let providers: [ProcessedDependencyProvider]
 
-    private func serialize(_ provider: ProcessedDependencyProvider) -> SerializedProvider {
-        let content = serializedContent(for: provider)
-        let registration = DependencyProviderRegistrationSerializer(provider: provider).serialize()
-        return SerializedProvider(content: content, registration: registration)
+    private func serialize(_ providers: [ProcessedDependencyProvider]) -> [SerializedProvider] {
+        var result = [SerializedProvider]()
+        let (baseClass, content) = serializedBase(for: providers.first!)
+        if !providers.first!.isEmptyDependency {
```

Comment:
Should we avoid this `!`, and do `== false` instead?

Reply:
done